README.md
# Data Flow

Data Flow is a custom ETL tool that uses airflow to manage data pipelines between different systems.


## Pipelines

There are currently 2 main types of pipelines in Data Flow:

* Data import pipelines fetch data from an external source (like a Data Hub or ONS API) and load them into a new dataset DB table.
* CSV pipelines query the dataset DB tables and generate a CSV file containing a view / cut of the data.

### Data import pipelines

As a result of a data import pipeline run a table in the datasets DB containing the source data is created (or replaced).
Data import pipelines should subclass the base `dataflow.dags._PipelineDAG` class and override `get_fetch_operator` method to return a fetch operator for a given data source.

All data import pipelines currently follow the same process:

1. Create a unique temporary DB table for the current task run.
2. Fetch data from the source and store responses as JSON lists in S3 bucket, prefixed with a unique string for each task run. This task is unique for each data source (e.g. Data Hub, ONS, Activity Stream), but should use the common `dataflow.utils.S3Data` class to make sure the stored responses can be accessed by the following DB tasks.
3. Read response files from S3 and import them into the temporary DB table.
4. Run checks on the new dataset. Currently this includes things like making sure the new dataset isn't significantly smaller than the previous version and that all DB columns are used.
5. Replace the current `.table_name` dataset table and the new table containing latest data.
6. Clean up any temporary / previous dataset tables at the end of the run.

### CSV pipelines

CSV pipelines run a specified query on existing dataset DB tables and save the query results in a CSV file stored in S3. The CSV files can then be downloaded through the Data Workspace.
CSV pipelines should subclass the base `dataflow.dags._CSVPipelineDAG` class and set the required values for `base_file_name` and `query` attributes.


## Running it with Docker

To run this in a local development environment, you'll need to get the dev environment's S3 service key from GOV.UK PaaS (https://www.cloud.service.gov.uk/). Set up an account, then ask webops for access to the `datasci-dev` space.

1. Copy sample.env to .env
2. Using the CF CLI (for GOV.UK PaaS), read data-flow's S3 service key and use it to populate the `AIRFLOW_CONN_DEFAULT_S3` and `S3_IMPORT_DATA_BUCKET`  env variables.
2. Run docker-compose up --build
3. Data-flow will be available on http://localhost:8080


## Deployment to production

Deployment is done via the jenkins "Data Flow" task by promoting a commit through dev, staging and production steps.


## Useful Information

- Airflow runs on UTC timezone by the community to prevent confusion that's why UI values are displayed in UTC timezone.
- Airflow creates a dag run for each completed interval between start date and end date. And it doesn't support scheduling tasks for end of each month. It means when August 1 2019 task is triggered, execution day is August 1 but current day is September 1 2019. Mind this when you filter your views. Please refer bellow code to get last day of execution month in jinja templated fields if you ever need it.
```
date_trunc('month', to_date('{{ macros.datetime.strptime(ds, '%Y-%m-%d') +
	macros.dateutil.relativedelta.relativedelta(months=+1, days=-1) }}', 'YYYY-MM-DD'));
```
- Logs show up after task completed it's the expected behaviour with Airflow remote logging. (S3 is used to centralized logs)

### Rerunning failed tasks

The pipelines are structured to make it possible to recover in case one of the tasks fails. In order to rerun a failed task you need to select it in the Airflow graph/tree view and click "Clear". This makes Airflow "forget" the previous result and reschedule the task. By default, Airflow will also clear any tasks that depend on the selected one ("downstream"), which is what we want most of the time.

Since things like temporary DB table names and S3 fetch cache locations are using DAG run ID, using "Clear" instead of triggering a new run means tasks will be able to reuse them. It also means any external task sensors that are still waiting will pick up the new result (built-in sensors only check the exact execution time, so will ignore any manually triggered DAG runs).

For data import pipelines, it's important to keep the state of DB tables in mind when choosing a task to re-run:

* If a pipeline fails during fetch or DB insert, the temporary table will get removed during DAG cleanup. So in order to rerun it, "create-temp-tables" task needs to be cleared first. If the fetch was successful it will rerun insert task using existing S3 cache, otherwise the entire DAG needs to be cleared in order to rerun fetch as well.
* If a check or swap DB tables tasks fail then the temporary table is left in place, so (as long as there were no bugs in fetch/insert tasks that needed a new release) only the failed tasks need to be cleared for a rerun.


### Managing Data-Flow migrations

Data-Flow maintains some models in the `dataflow` schema of the datasets DB. At the moment this is just a `metadata` table used to associate some timestamps with tables. Migrations for this table, and future tables in the same schema, are managed using Alembic. To change the model:

1) Update the model in Python
2) Run `PYTHONPATH=. alembic revision --autogenerate -m "<short description>"`
3) Optionally run `PYTHONPATH=. alembic upgrade head`, though this is run automatically on app startup.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## 2020-10-28

### Added

- Pipeline for DIT zendesk tickets (ZendeskDITTradeTicketsPipeline)
- Pipeline for UKTrade zendesk tickets (ZendeskUKTradeTicketsPipeline)

## 2020-10-26

### Added

- New pipeline for companies house significant persons of control

## 2020-10-22

### Changed

- Rename and move LITE to tables to dit schema 
- Use the new "latest" postcodes materialized view in dashboards

## 2020-10-21

### Changed

- Add `export_revenue` field to investment projects datashboard pipeline

### Added

- New, cabinet office gender pay gap pipeline

## 2020-10-20

### Changed

- More retrying and longer waits between retries for Activity Stream pipelines

## 2020-10-16

### Added

- Add 5 new fields to the market access dataset
    - trading_bloc_code
    - trading_bloc_name
    - end_date
    - summary
    - wto_profile

## 2020-10-15

### Changed

- Update ONSPostcodesPipeline to keep data for all publication dates and add a view per publication

## 2020-10-14

### Changed

- Split the requests for sharepoint columns/data to allow for pagination  

## 2020-10-13

### Changed

- Retry all _DatasetPipeline 3 times to work around them being killed somewhat randomly
- Ensure hmrc pipelines include all available historic data

## 2020-10-05

### Added
- Add [prometheus metrics exporter](https://github.com/uktrade/data-flow-metrics) to data-flow.

### Changed
- Set default retries across almost all DAGs at three, to (hopefully) reduce the need for manually restarting failed pipelines that are caused by transient issues.

## 2020-10-02

### Added

- Pipeline for interaction tags classifier (TagsClassifierPipeline).

### Changed

- More export wins derived fetch task to the high mem queue

## 2020-09-30

### Changed

- Update country code mapping for the JHU covid dataset

## 2020-09-29

### Changed

- Join on original rather than transformed sectors for DataHubMonthlyInvesmentProjectsPipline query

## 2020-09-24

### Added

- Maxemail campaigns sent emails pipeline.
- Email alerts for the remaining 3 ONS pipeline runs.

### Changed

- Preserve views/materialised views dependencies on table swap

## 2020-09-23

### Added

- New activity stream pipeline for the return-to-office service

### Changed

- Fix typos in export wins derived config
- Fix incorrect field type bug in return to office pipeline
- Add correct id field to return to office pipeline
- Use matching service update endpoint for some matching pipeline 

## 2020-09-22

### Added

- HMRC Exporters matching pipeline

### Changed

- Support Array types in DSSGenericPipeline

## 2020-09-21

### Changed

- Run the google covid pipeline daily 
- Allow null columns on the derived export wins pipeline
- Use the high memory queue for the google covid dataset fetch operation
- Add region name and region code fields to oxford covid dataset

## 2020-09-18

### Added

- Add a new derived table for reporting on export wins

## 2020-09-15

### Changed
- Use postgres's native `copy_from` rather than `DataFrame.to_sql` to load data in postgres for the "fast polling" pipelines for significant speed gains (~15x).

## 2020-09-14

### Added

- Contact and advisor last interaction pipelines.

### Changed
- ONS UK trade in services migrated to new polling pipeline.
- ONS UK total trade migrated to new polling pipeline.

## 2020-09-12

### Added

- The ability to send notification emails to users when a pipeline has completed successfully, configured through env vars.

### Changed

- ONS UK SA Trade in Goods dataset to scrape/clean/transform/load directly from the ONS spreadsheet rather than their alpha API, which has been flaky. This will also use our condensed polling pipeline to load this data more quickly.
- ONS UK Trade in goods by country and commodity dataset to a condensed polling pipeline.

## 2020-09-11

### Added

- Allow for pipelines to specify setting `source_data_modified_utc` to run time (utc)

### Changed

- Update daily dataset dags to set `source_data_modified_utc` to utc now

## 2020-09-08

### Changed

- Update queries using legacy `ref_countries_and_territories` reference datasets to use new `ref_countries_territories_and_regions` dataset

## 2020-09-04

### Added

- HMRC Field Force matching pipeline

## 2020-09-03

### Changed

- Fix incorrect column name in People Finder people pipeline

## 2020-09-01

### Added

- 'country_investment_originates_from' to Investment Projects dataset.

## 2020-08-28

### Added

- New investment projects pipeline DataHubMonthlyInvesmentProjectsPipline

## 2020-08-26

### Changed

- Updated fields in People Finder people pipeline

## 2020-08-25

### Added

- 'SSO email' to Advisers dataset.

## 2020-08-24

### Changed

- Add numeric data type to generic dss pipeline

## 2020-08-24

### Changed

- Renamed DirectoryFormsPipeline to GreatGovUKFormsPipeline.

## 2020-08-21

### Changed

- Update maintenance dag to include all (non user and postgres) schemas

## 2020-08-20

### Added

- Directory Forms API pipeline.
- New HMRC pipeline for exports to the EU
- New HMRC pipeline for import from the EU 

### Changed

- Use hmrc schema for all hmrc pipelines

## 2020-08-18

### Added

- Metadata table to the datasets DB under a `dataflow` schema, used to track various information related to pipeline runs/data pulled in. To begin with, we record the last modification date for ONS datasets and the timestamp when dataflow swapped temporary ingest tables with the "real" tables.
- Updated ONS parsing pipelines to only continue if there is more recent data available, based on the source data modified timestamp stored in the metadata table.

## 2020-08-17

### Added

- New pipeline to sync changes from the dnb company service to the dnb global companies table

## 2020-08-13

### Changed


dataflow/dags/base.py
import sys

from datetime import datetime, timedelta, time
from functools import partial
from typing import List, Optional, Type, Callable, Tuple

from airflow import DAG
from airflow.models import SkipMixin
from airflow.operators.dummy_operator import DummyOperator
from airflow.operators.python_operator import PythonOperator, BranchPythonOperator
from airflow.operators.sensors import ExternalTaskSensor

from dataflow import config
from dataflow.operators.csv_outputs import create_csv, create_compressed_csv
from dataflow.operators.db_tables import (
    check_table_data,
    create_temp_tables,
    drop_swap_tables,
    drop_temp_tables,
    insert_data_into_db,
    swap_dataset_tables,
    branch_on_modified_date,
    scrape_load_and_check_data,
    poll_for_new_data,
)
from dataflow.operators.email import send_dataset_update_emails
from dataflow.utils import TableConfig, slack_alert, SingleTableConfig


class PipelineMeta(type):
    """Metaclass to register pipeline DAGs for airflow discovery.

    When creating a class object also creates an instance of the class and
    saves the `.get_dag()` return value in the class module's __dict__,
    where it can be discovered by Airflow.

    To avoid registering DAGs for base classes, this skips a class if its
    name starts with `_`.

    """

    def __new__(mcls, name, bases, attrs):
        pipeline = super(PipelineMeta, mcls).__new__(mcls, name, bases, attrs)

        if not pipeline.__name__.startswith('_'):
            sys.modules[pipeline.__module__].__dict__[
                pipeline.__name__ + "__dag"
            ] = pipeline().get_dag()

        return pipeline


class _PipelineDAG(metaclass=PipelineMeta):
    target_db: str = config.DATASETS_DB_NAME
    start_date: datetime = datetime(2019, 11, 5)
    end_date: Optional[datetime] = None
    schedule_interval: str = "@daily"
    catchup: bool = False

    # Enable or disable Slack notification when DAG run finishes
    alert_on_success: bool = False
    alert_on_failure: bool = True

    # Disables the null columns check that makes sure all DB columns
    # have at least one non-null value
    allow_null_columns: bool = False

    table_config: TableConfig

    dependencies: List[Type["_PipelineDAG"]] = []

    # Controls how long will the task wait for dependencies to succeed
    dependencies_timeout: int = 8 * 3600

    # Offset between dependencies and this DAG running. So for example
    # if this pipeline starts at 5am and depends on a task that starts
    # at midnight the offset should be 5 hours. This will check for the
    # exact time, NOT a range of "between now and 5 hours ago".
    dependencies_execution_delta: timedelta = timedelta(hours=0)

    # If True, the source_data_modified_utc metadata field will be set to the swap table task run utc.
    use_utc_now_as_source_modified: bool = False

    def get_fetch_operator(self) -> PythonOperator:
        raise NotImplementedError(
            f"{self.__class__} needs to override get_fetch_operator"
        )

    def get_insert_data_callable(self):
        return insert_data_into_db

    def get_source_data_modified_utc(self) -> Optional[datetime]:
        return None

    def get_source_data_modified_utc_callable(self) -> Optional[Callable]:
        return None

    def branch_on_modified_date(
        self, dag: DAG, target_db: str, table_config: TableConfig
    ) -> PythonOperator:
        """Check whether data is newer than the previous run, else abort.

        This task is executed if the DAG returns non-None from `get_source_data_modified_utc_callable` method. That
        method should return a callable which itself returns a UTC timestamp that is compared gainst the
        `dataflow.metadata` table for this pipeline. If the date is newer than our stored date, the pipeline will
        continue, otherwise we'll abort.
        """
        return BranchPythonOperator(
            task_id="branch-on-modified-date",
            python_callable=branch_on_modified_date,
            execution_timeout=timedelta(minutes=10),
            provide_context=True,
            op_args=[target_db, table_config],
            dag=dag,
        )

    def get_transform_operator(self):
        """
        Optional overridable task to transform/manipulate data
        between insert-into-temp-table task and check-temp-table-data task
        """
        return None

    @classmethod
    def fq_table_name(cls):
        return f'"{cls.table_config.schema}"."{cls.table_config.table_name}"'

    def get_dag(self) -> DAG:
        dag = DAG(
            self.__class__.__name__,
            catchup=self.catchup,
            default_args={
                "owner": "airflow",
                "depends_on_past": False,
                "email_on_failure": False,
                "email_on_retry": False,
                "retries": 3,
                "retry_delay": timedelta(minutes=5),
                'catchup': self.catchup,
            },
            start_date=self.start_date,
            end_date=self.end_date,
            schedule_interval=self.schedule_interval,
            max_active_runs=1,
            on_success_callback=partial(slack_alert, success=True)
            if self.alert_on_success
            else None,
            on_failure_callback=partial(slack_alert, success=False)
            if self.alert_on_failure
            else None,
        )

        # If we've configured a way for the pipeline to determine when the source data was last modified,
        # then we should run checks to see if the data has been updated. If it hasn't, we will end the pipeline
        # early (but gracefully) without pulling the data again.
        _get_source_data_modified_utc_callable = (
            self.get_source_data_modified_utc_callable()
        )
        if _get_source_data_modified_utc_callable:
            _get_source_modified_date_utc = PythonOperator(
                task_id="get-source-modified-date",
                python_callable=_get_source_data_modified_utc_callable,
                dag=dag,
            )
            _branch_on_modified_date = self.branch_on_modified_date(
                dag, self.target_db, self.table_config
            )

            _stop = DummyOperator(task_id="stop", dag=dag)
            _continue = DummyOperator(task_id="continue", dag=dag)
        else:
            _branch_on_modified_date = None
            _get_source_modified_date_utc = None
            _stop = None
            _continue = None

        _fetch = self.get_fetch_operator()
        _fetch.dag = dag

        _create_tables = PythonOperator(
            task_id="create-temp-tables",
            python_callable=create_temp_tables,
            execution_timeout=timedelta(minutes=10),
            provide_context=True,
            op_args=[self.target_db, *self.table_config.tables],
            dag=dag,
        )

        _insert_into_temp_table = PythonOperator(
            task_id="insert-into-temp-table",
            python_callable=self.get_insert_data_callable(),
            provide_context=True,
            op_kwargs=(dict(target_db=self.target_db, table_config=self.table_config)),
        )

        _transform_tables = self.get_transform_operator()
        if _transform_tables:
            _transform_tables.dag = dag

        _check_tables = PythonOperator(
            task_id="check-temp-table-data",
            python_callable=check_table_data,
            provide_context=True,
            op_args=[self.target_db, *self.table_config.tables],
            op_kwargs={'allow_null_columns': self.allow_null_columns},
        )

        _swap_dataset_tables = PythonOperator(
            task_id="swap-dataset-table",
            python_callable=swap_dataset_tables,
            execution_timeout=timedelta(minutes=10),
            provide_context=True,
            op_args=[self.target_db, *self.table_config.tables],
            op_kwargs={
                'use_utc_now_as_source_modified': self.use_utc_now_as_source_modified
            },
        )

        _drop_temp_tables = PythonOperator(
            task_id="drop-temp-tables",
            python_callable=drop_temp_tables,
            execution_timeout=timedelta(minutes=10),
            provide_context=True,
            trigger_rule="one_failed",
            op_args=[self.target_db, *self.table_config.tables],
        )

        _drop_swap_tables = PythonOperator(
            task_id="drop-swap-tables",
            python_callable=drop_swap_tables,
            execution_timeout=timedelta(minutes=10),
            provide_context=True,
            op_args=[self.target_db, *self.table_config.tables],
        )

        if _get_source_modified_date_utc:
            _get_source_modified_date_utc >> _branch_on_modified_date >> [
                _stop,
                _continue,
            ]
            _continue >> [_fetch, _create_tables]

        [_fetch, _create_tables] >> _insert_into_temp_table >> _drop_temp_tables

        if _transform_tables:
            _insert_into_temp_table >> _transform_tables >> _check_tables
        else:
            _insert_into_temp_table >> _check_tables

        _check_tables >> _swap_dataset_tables >> _drop_swap_tables

        for dependency in self.dependencies:
            sensor = ExternalTaskSensor(
                task_id=f'wait-for-{dependency.__name__.lower()}',
                pool="sensors",
                external_dag_id=dependency.__name__,
                external_task_id='swap-dataset-table',
                execution_delta=self.dependencies_execution_delta,
                timeout=self.dependencies_timeout,
                dag=dag,
            )

            sensor >> [_fetch, _create_tables]

        return dag


class _CSVPipelineDAG(metaclass=PipelineMeta):
    target_db: str = config.DATASETS_DB_NAME
    start_date: datetime = datetime(2019, 11, 5)
    end_date: Optional[datetime] = None
    schedule_interval: str = "@daily"
    catchup: bool = True

    # Static pipeliens are not refreshed daily, so generated files won't be updated
    static: bool = False

    # S3 file name prefix
    base_file_name: str

    # Controls whether the current timestamp is appended to the base_file_name in S3 path
    timestamp_output: bool = True

    # DB query to generate data for the CSV file
    query: str

    # Should the output CSV be compressed (zip archive) before uploading to S3.
    compress: bool = False

    dependencies: List[Type["_PipelineDAG"]] = []

    # Controls how long will the task wait for dependencies to succeed
    dependencies_timeout: int = 8 * 3600

    # Offset between dependencies and this DAG running. So for example
    # if this pipeline starts at 5am and depends on a task that starts
    # at midnight the offset should be 5 hours. This will check for the
    # exact time, NOT a range of "between now and 5 hours ago".
    dependencies_execution_delta: timedelta = timedelta(hours=0)

    def get_dag(self) -> DAG:
        dag = DAG(
            self.__class__.__name__,
            catchup=self.catchup,
            default_args={
                'owner': 'airflow',
                'depends_on_past': False,
                'email_on_failure': False,
                'email_on_retry': False,
                'retries': 3,
                'retry_delay': timedelta(minutes=5),
                'start_date': datetime(2019, 1, 1),
            },
            user_defined_macros={"dependencies": self.dependencies},
            start_date=self.start_date,
            end_date=self.end_date,
            schedule_interval=self.schedule_interval,
        )

        _create_csv = PythonOperator(
            task_id='create-csv-current',
            python_callable=create_compressed_csv if self.compress else create_csv,
            provide_context=True,
            op_args=[
                self.target_db,
                self.base_file_name,
                self.timestamp_output,
                self.query,
            ],
            dag=dag,
        )

        for dependency in self.dependencies:
            sensor = ExternalTaskSensor(
                task_id=f'wait-for-{dependency.__name__.lower()}',
                pool="sensors",
                external_dag_id=dependency.__name__,
                external_task_id='swap-dataset-table',
                execution_delta=self.dependencies_execution_delta,
                timeout=self.dependencies_timeout,
                dag=dag,
            )

            sensor >> [_create_csv]

        return dag


class _FastPollingPipeline(SkipMixin, metaclass=PipelineMeta):
    """
    A pipeline that continuously polls (within a given period) for updates to a dataset and then aims to process
    and upload that data as fast as practicable, to support more time-sensitive workflows. The pipeline minimises time
    by skipping or condensing steps that are in the standard `_PipelineDAG`, e.g. uploading data to S3 and splitting
    steps out into separate tasks.
    """

    target_db: str = config.DATASETS_DB_NAME
    start_date: datetime = datetime(2019, 11, 5)
    end_date: Optional[datetime] = None
    catchup: bool = False

    # Enable or disable Slack notification when DAG run finishes
    alert_on_success: bool = False
    alert_on_failure: bool = True

    # Disables the null columns check that makes sure all DB columns
    # have at least one non-null value
    allow_null_columns: bool = False

    # These two functions must be defined on any subclasses
    date_checker: Callable[[], Tuple[datetime, Optional[datetime]]]
    data_getter: Callable

    table_config: SingleTableConfig

    # How often to poll the data source to read it's "last modified" date.
    polling_interval_in_seconds = 60

    schedule_interval = "0 6 * * *"
    daily_end_time_utc = time(17, 0, 0)

    # Which worker to run the poll/scrape/clean/load task on.
    worker_queue = 'default'

    # If this is defined, it should point to an environment variable that provides a small blob of JSON data:
    # {
    #   "dataset_name": "A friendly name for the dataset",
    #   "dataset_url": "https://www.data.trade.gov.uk/datasets/...",
    #   "emails": ["subscriber@data.trade.gov.uk", ...]
    # }
    update_emails_data_environment_variable: Optional[str] = None

    @classmethod
    def fq_table_name(cls):
        return f'"{cls.table_config.schema}"."{cls.table_config.table_name}"'

    def skip_downstream_tasks(self, **kwargs):
        downstream_tasks = kwargs['task'].get_flat_relatives(upstream=False)
        self.skip(kwargs['dag_run'], kwargs['ti'].execution_date, downstream_tasks)

    def get_dag(self) -> DAG:
        dag = DAG(
            self.__class__.__name__,
            catchup=self.catchup,
            default_args={
                "owner": "airflow",
                "depends_on_past": False,
                "email_on_failure": False,
                "email_on_retry": False,
                "retries": 3,
                "retry_delay": timedelta(seconds=30),
                'catchup': self.catchup,
            },
            start_date=self.start_date,
            end_date=self.end_date,
            schedule_interval=self.schedule_interval,
            max_active_runs=1,
            on_success_callback=partial(slack_alert, success=True)
            if self.alert_on_success
            else None,
            on_failure_callback=partial(slack_alert, success=False)
            if self.alert_on_failure
            else None,
        )

        _poll_for_updates = PythonOperator(
            task_id='poll-for-new-data',
            python_callable=poll_for_new_data,
            op_kwargs=dict(
                target_db=self.target_db,
                table_config=self.table_config,
                pipeline_instance=self,
            ),
            dag=dag,
            provide_context=True,
        )

        _scrape_load_and_check = PythonOperator(
            task_id='scrape-and-load-data',
            python_callable=scrape_load_and_check_data,
            op_kwargs=dict(
                target_db=self.target_db,
                table_config=self.table_config,
                pipeline_instance=self,
            ),
            dag=dag,
            provide_context=True,
            queue=self.worker_queue,
        )

        _swap_dataset_tables = PythonOperator(
            task_id="swap-dataset-table",
            dag=dag,
            python_callable=swap_dataset_tables,
            execution_timeout=timedelta(minutes=10),
            provide_context=True,
            op_args=[self.target_db, *self.table_config.tables],
        )

        _send_dataset_updated_emails = None
        if self.update_emails_data_environment_variable:
            _send_dataset_updated_emails = PythonOperator(
                task_id='send-dataset-updated-emails',
                dag=dag,
                retries=0,
                python_callable=send_dataset_update_emails,
                op_kwargs=dict(
                    update_emails_data_environment_variable=self.update_emails_data_environment_variable
                ),
            )

        _drop_temp_tables = PythonOperator(
            task_id="drop-temp-tables",
            dag=dag,
            python_callable=drop_temp_tables,
            execution_timeout=timedelta(minutes=10),
            provide_context=True,
            trigger_rule="one_failed",
            op_args=[self.target_db, *self.table_config.tables],
        )

        _drop_swap_tables = PythonOperator(
            task_id="drop-swap-tables",
            dag=dag,
            python_callable=drop_swap_tables,
            execution_timeout=timedelta(minutes=10),
            provide_context=True,
            op_args=[self.target_db, *self.table_config.tables],
        )

        (
            _poll_for_updates
            >> _scrape_load_and_check
            >> _swap_dataset_tables
            >> [_drop_swap_tables, _drop_temp_tables]
        )

        if _send_dataset_updated_emails:
            _swap_dataset_tables >> _send_dataset_updated_emails

        return dag
